### DISCUSSION TITLE
Installation process

### DISCUSSION THREAD

**Discussion by vladimircezar** [danielmiessler/Fabric#87] [Q&A]
The client installation process is not working for me. 
After trying to go through the Quickstart (and verifying the $PATH) I keep receiving `zsh: no such file or directory` 

Then I tried using the READ.ME (which contains a slightly different set of steps) and still could not make the client run.

Can someone give a hint on what needs to be done?

**Comment by protteveel** [comment]
In the section **Installing the server**

The command:
```
echo 'alias fabric-server="/the/path/to/fabric/server/fabric-api"' >> ~/.bashrc
```
Should probably state:
```
echo 'alias fabric-api="/the/path/to/fabric/server/fabric-api"' >> ~/.bashrc
```
The text:

> PS: fabric-api and fabric-webui need to be run in dictinct terminal windows.

Should probably state:

> PS: fabric-api and fabric-webui need to be run in distinct terminal windows.

In the section **Create your own Fabric Mill**

The file /client/standalone_client_examples does not exist.

The link [/client/standalone_client_examples](https://github.com/danielmiessler/fabric/tree/main/client/standalone_client_examples) does not work.

**Comment by protteveel** [comment]
I have fabric-api up-and-running on http://127.0.0.1:13337

1. _How do I interface with it?_
2. _How and where do I add patterns?_

I have fabric-webui up-and-running on http://127.0.0.1:13338

1. _How and where do I add patterns?_
2. _How do I solve the following error message?_

If I use the web interface for fabric-webui, with "This is a test" for **Content** and use "/extwis" for **API**, I get the following error messages when I hit **Send Request**:
```
Traceback (most recent call last):
  File "/Users/percyrotteveel/Library/Caches/pypoetry/virtualenvs/fabric-server-tG7n8Hc0-py3.11/lib/python3.11/site-packages/flask/app.py", line 1488, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/percyrotteveel/Library/Caches/pypoetry/virtualenvs/fabric-server-tG7n8Hc0-py3.11/lib/python3.11/site-packages/flask/app.py", line 1466, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/percyrotteveel/Library/Caches/pypoetry/virtualenvs/fabric-server-tG7n8Hc0-py3.11/lib/python3.11/site-packages/flask/app.py", line 1463, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/percyrotteveel/Library/Caches/pypoetry/virtualenvs/fabric-server-tG7n8Hc0-py3.11/lib/python3.11/site-packages/flask/app.py", line 872, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/percyrotteveel/Library/Caches/pypoetry/virtualenvs/fabric-server-tG7n8Hc0-py3.11/lib/python3.11/site-packages/flask/app.py", line 870, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/percyrotteveel/Library/Caches/pypoetry/virtualenvs/fabric-server-tG7n8Hc0-py3.11/lib/python3.11/site-packages/flask/app.py", line 855, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/percyrotteveel/workspace/fabric/server/fabric_server/fabric_web_interface/fabric_web_server.py", line 84, in index
    response = send_request(prompt=prompt, endpoint=endpoint)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/percyrotteveel/workspace/fabric/server/fabric_server/fabric_web_interface/fabric_web_server.py", line 36, in send_request
    "Authorization": f"Bearer {session['token']}",
                               ^^^^^^^^^^^^^^^^^^^
  File "/Users/percyrotteveel/Library/Caches/pypoetry/virtualenvs/fabric-server-tG7n8Hc0-py3.11/lib/python3.11/site-packages/flask/sessions.py", line 84, in __getitem__
    return super().__getitem__(key)
           ^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'token'
```

**Reply by Foxur** [reply]
Any updates for the KeyError? My session seems not existing <NullSession {}>.
Any other option to make an Authentication manuly by hardcoding a token for the session?

**Comment by danielmiessler** [comment]
So you have to use the standalone fabric client to hit these, and the patterns are in the server code itself. They're hardcoded into the server. So that's where you edit them.

**Reply by protteveel** [reply]
In the file fabric_api_server.py, I see the following patterns:
```python
pattern_path_mappings = {
    "extwis": {"system_url": "https://raw.githubusercontent.com/danielmiessler/fabric/main/patterns/extract_wisdom/system.md",
               "user_url": "https://raw.githubusercontent.com/danielmiessler/fabric/main/patterns/extract_wisdom/user.md"},
    "summarize": {"system_url": "https://raw.githubusercontent.com/danielmiessler/fabric/main/patterns/summarize/system.md",
                  "user_url": "https://raw.githubusercontent.com/danielmiessler/fabric/main/patterns/summarize/user.md"}
} # Add more pattern with your desire path as a key in this dictionary
```
**![Screenshot 2024-02-27 at 09 38 00](https://github.com/danielmiessler/fabric/assets/5226789/527b9cff-5e76-48d2-b3a4-fcc9ec4ad4e0)**
1. However, in the Web Gui, it only shows the extwis pattern; not the summarize pattern. How to fix this?
2. Additionally, I would like to use the patterns on my local machine, provided by my own fabric server. Can you explain how to do that?

**Reply by nic0711** [reply]
Maybe we could share the /pattern folder across the system?
Redundant, hard coded patterns seems hard to maintain and update. Just my 5cts.

**Reply by jmatthewpryor** [reply]
It also seems like there is some login/register protocol on the server (register a user, login with a username/password) to set a JWT token, but the web ui code seems to fail because there is no JWT?

**Comment by nic0711** [comment]
Poetry recommends the use of a dedicated env...

I use conda for different envs, so fabric will not be available without changing the env. 

what are your experience's?
